tools/scripts/seed-exams/exam-schema.js

```diff
@@ -0,0 +1,101 @@
+import Joi from 'joi';
+import JoiObjectId from 'joi-objectid';
+
+Joi.objectId = JoiObjectId(Joi);
+
+const nanoIdRE = new RegExp('[a-z0-9]{10}');
+
+const PrerequisitesJoi = Joi.object().keys({
+  id: Joi.objectId().required(),
+  title: Joi.string()
+});
+
+const AnswerJoi = Joi.object().keys({
+  id: Joi.string().regex(nanoIdRE).required(),
+  deprecated: Joi.bool(),
+  answer: Joi.string().required()
+});
+
+const QuestionJoi = Joi.object().keys({
+  id: Joi.string().regex(nanoIdRE).required(),
+  question: Joi.string().required(),
+  deprecated: Joi.bool(),
+  wrongAnswers: Joi.array()
+    .items(AnswerJoi)
+    .min(4)
+    .required()
+    .custom((value, helpers) => {
+      const deprecatedCount = value.reduce(
+        (count, answer) => (answer.deprecated === true ? count + 1 : count),
+        0
+      );
+      const minimumAnswers = 4;
+      const nonDeprecatedCount = value.length - deprecatedCount;
+
+      if (nonDeprecatedCount < minimumAnswers) {
+        return helpers.message(
+          `'wrongAnswers' must have at least ${minimumAnswers} non-deprecated answers.`
+        );
+      }
+
+      return value;
+    }),
+  correctAnswers: Joi.array()
+    .items(AnswerJoi)
+    .min(1)
+    .required()
+    .custom((value, helpers) => {
+      const deprecatedCount = value.reduce(
+        (count, answer) => (answer.deprecated === true ? count + 1 : count),
+        0
+      );
+      const minimumAnswers = 1;
+      const nonDeprecatedCount = value.length - deprecatedCount;
+
+      if (nonDeprecatedCount < minimumAnswers) {
+        return helpers.message(
+          `'correctAnswers' must have at least ${minimumAnswers} non-deprecated answer.`
+        );
+      }
+
+      return value;
+    })
+});
+
+const schema = Joi.object()
+  .keys({
+    _id: Joi.objectId().required(),
+    title: Joi.string().required(),
+    numberOfQuestionsInExam: Joi.number().min(0).max(10).required(),
+    passingPercent: Joi.number().min(0).max(100).required(),
+    prerequisites: Joi.array().items(PrerequisitesJoi),
+    questions: Joi.array().items(QuestionJoi).min(1).required()
+  })
+  .when(
+    Joi.object({
+      numberOfQuestionsInExam: Joi.number().required(),
+      questions: Joi.array().required()
+    }).unknown(),
+    {
+      then: Joi.object({
+        numberOfQuestionsInExam: Joi.number()
+          .max(
+            Joi.ref('questions', {
+              adjust: questions => {
+                const deprecatedQuestionsCount = questions.reduce(
+                  (count, question) =>
+                    question.deprecated === true ? count + 1 : count,
+                  0
+                );
+                return questions.length - deprecatedQuestionsCount;
+              }
+            })
+          )
+          .required()
+      })
+    }
+  );
```

Comment:
```suggestion
const schema = Joi.object().keys({
  _id: Joi.objectId().required(),
  title: Joi.string().required(),
  numberOfQuestionsInExam: Joi.number()
    .min(0)
    .max(
      Joi.ref('questions', {
        adjust: questions => {
          const nonDeprecatedCount = questions.reduce(
            (count, question) => (question.deprecated ? count : count + 1),
            0
          );
          return nonDeprecatedCount < 10 ? nonDeprecatedCount : 10;
        }
      })
    )
    .required(),
  passingPercent: Joi.number().min(0).max(100).required(),
  prerequisites: Joi.array().items(PrerequisitesJoi),
  questions: Joi.array().items(QuestionJoi).min(1).required()
});
```
I could have missed the point, but isn't this schema supposed to cap `numberOfQuestionsInExam` at either 10 or the number of questions, whichever is smallest?

If so, I think this makes that more obvious.

Reply:
It's not supposed to be 10, that's a mistake. It's supposed to be `numberOfQuestionsInExam <= nonDeprecatedQuestions.length`

Reply:
The `fix: validate number of questions better` should resolve this.